app/src/main/java/com/zulip/android/activities/ZulipActivity.java

```diff
@@ -535,6 +540,142 @@ public Cursor runQuery(CharSequence charSequence) {
+                if (grantResults.length > 0
+                        && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                    // permission granted
+                    // start with file upload
+                    startFileUpload();
+                } else {
+                    // permission denied
+                    Toast.makeText(this, R.string.cannot_upload_image, Toast.LENGTH_SHORT).show();
+                }
+            }
+            break;
+        }
+    }
+
+    /**
+     * Helper function to update UI to indicate image is being uploaded and call
+     * {@link ZulipActivity#uploadFile(String)} to upload the image.
+     */
+    private void startFileUpload() {
+        // Update UI to indicate image is being loaded
+        // hide fab and display chatbox
+        displayFAB(false);
+        displayChatBox(true);
+        String loadingMsg = getResources().getString(R.string.uploading_message);
+        sendingMessage(true, loadingMsg);
+
+        // get actual file path
+        String imageFilePath = FilePathHelper.getPath(this, mImageUri);
+
+        // upload the file asynchronously to the server
+        uploadFile(imageFilePath);
+    }
+
+    /**
+     * Function to upload file asynchronously to the server using retrofit callback
+     * upload {@link com.zulip.android.service.ZulipServices#upload(MultipartBody.Part)}
+     * @param filePath on local storage
+     */
+    private void uploadFile(String filePath) {
+        File file = new File(filePath);
+
+        // create RequestBody instance from file
+        RequestBody requestFile =
+                RequestBody.create(MediaType.parse("multipart/form-data"), file);
+
+        // MultipartBody.Part is used to send also the actual file name
+        MultipartBody.Part body =
+                MultipartBody.Part.createFormData("picture", file.getName(), requestFile);
+
+        final String loadingMsg = getResources().getString(R.string.uploading_message);
+
+        // finally, execute the request
+        // create upload service client
+        Call<UploadResponse> call = ((ZulipApp) getApplicationContext()).getZulipServices().upload(body);
+        call.enqueue(new Callback<UploadResponse>() {
+            @Override
+            public void onResponse(Call<UploadResponse> call,
+                                   Response<UploadResponse> response) {
+                if (response.isSuccessful()) {
+                    String filePathOnServer = "";
+                        UploadResponse uploadResponse = response.body();
+                        filePathOnServer = uploadResponse.getUri();
+                    if (!filePathOnServer.equals("")) {
+                        // remove loading message from the screen
+                        sendingMessage(false, loadingMsg);
+
+                        // print message to compose box
+                        messageEt.append(" " + UrlHelper.addHost(filePathOnServer));
```

Comment:
This is a nit pick, but the urls that this is generating have two slashes in them at the seam between the host and the path: ie `https://zulip.tabbott.net//user_uploads/2/3c/lpFDn37AU7VWyJg4sZruImIX/IMG_20161015_155048.jpg`